feat: Add project depends on functionality

Luay-Sol opened the pull request:
## what
- The main purpose of this PR is to add Dependencies functionality to Atlantis. The change introduces a new key `depends_on` to the project config in the `atlantis.yaml` file.

## why
Using the `depends_on` functionality will allow Atlantis to handle cases requiring one project to be applied prior to the other. In other words, allow promotions between projects/environments (dev -> staging -> production)
 
## tests
- [X] I have tested my changes by adding new unit tests, running Atlantis locally, and testing the depends_on functionality.  

## references
- Somehow related to this feature request: https://github.com/runatlantis/atlantis/issues/391, but not exactly. The implementation followed in this PR does not depend on the apply requirement key.

jamengual commented on server/core/config/valid/global_cfg.go:
```diff
@@ -17,10 +17,7 @@ const PoliciesPassedCommandReq = "policies_passed"
 const PlanRequirementsKey = "plan_requirements"
 const ApplyRequirementsKey = "apply_requirements"
 const ImportRequirementsKey = "import_requirements"
-const PreWorkflowHooksKey = "pre_workflow_hooks"
```
why are you deleting these constants?

Luay-Sol replied:
Looked at the code base and it seems they are not being used.

jamengual replied:
can you rever that, I believe those are being used.

GenPage replied:
They are not being used. It's mainly used in the raw/global_cfg.go to parse the config. 

See: https://github.com/runatlantis/atlantis/blob/main/server/core/config/raw/global_cfg.go#L194-L195 as an example

jamengual commented on server/events/command/project_context.go:
```diff
@@ -62,6 +69,8 @@ type ProjectContext struct {
 	PullReqStatus models.PullReqStatus
 	// CurrentProjectPlanStatus is the status of the current project prior to this command.
 	ProjectPlanStatus models.ProjectPlanStatus
+
```
could you add a comment here of what this is doing?

jamengual commented:
will these work by running `atlantis plan` and then `atlantis apply` in all projects and the apply will go though the dependencies in order?

jamengual commented:
Thanks for the contribution @Luay-Sol

Luay-Sol commented:
> will these work by running `atlantis plan` and then `atlantis apply` in all projects and the apply will go though the dependencies in order?

Actually, this feature is more for the case where the user try to apply to specific project/workspace. What you have described already exists in Atlantis and is called `execution_order_group`. So I see this feature will work in conjunction with the `execution_order_group`. For example this is how my project block look like:

```
projects:
  - name: development
    dir: .
    autoplan:
      when_modified: ["*.tf", "vars/development.tfvars"]
    execution_order_group: 1
    workspace: development
    workflow: infra
  - name:staging
    dir: .
    autoplan:
      when_modified: ["*.tf", "vars/staging.tfvars"]
    depends_on: ["development"]
    execution_order_group: 2
    workspace: staging
    workflow: infra
  - name: production
    dir: .
    autoplan:
      when_modified: ["*.tf", "vars/production.tfvars"]
    depends_on: ["staging"]
    execution_order_group: 3
    workspace: production
    workflow: infra
```

So with this, if the user just comments `atlantis apply` the `execution_order_group` will ensure that the changes will go `development->staging->production`. Else if the user makes the apply command sequentially   `atlantis apply -w development`, then `atlantis apply -w staging`, then `atlantis apply -w production`, the `depends_on` feature will make sure that production is not applied before staging for example.

jamengual commented:
Amazing that is what I was expecting and I can see how useful this will be.

GenPage commented on server/core/config/valid/repo_cfg.go:
```diff
@@ -129,6 +129,7 @@ type Project struct {
 	PlanRequirements          []string
 	ApplyRequirements         []string
 	ImportRequirements        []string
+	Dependencies              []string
```
Rename to `DependsOn`

GenPage commented on server/events/command/project_context.go:
```diff
@@ -54,6 +54,13 @@ type ProjectContext struct {
+	// Atlantis uses this information to valid the apply
+	// orders and to warn the user if they're applying a project that
+	// depends on other projects.
+	Dependencies []string
```
Rename to `DependsOn`

GenPage commented on server/events/command_requirement_handler_test.go:
```diff
@@ -194,6 +194,89 @@ func TestAggregateApplyRequirements_ValidateApplyProject(t *testing.T) {
+				},
+			},
+			wantFailure: "Can't apply your project unless you apply its dependencies: [project2]",
+			wantErr:     assert.NoError,
```
nitpick: Tests are good so far, can we include some more scenarios that include 3 projects to make sure Atlantis doesn't continue past the 2nd project?

Project 1: success
Project 2: fail
Project 3: skipped

vincentgna replied:
@Luay-Sol - re: the bug you were struggling with, is there a test case in here that ensures no regressions against that bug?

vincentgna replied:
oh, so the bug was related to missing 

```go
&& project.Status != models.PlannedNoChangesPlanStatus
```
to break the dependency chain

GenPage commented on server/events/command_requirement_handler.go:
```diff
@@ -65,6 +67,20 @@ func (a *DefaultCommandRequirementHandler) ValidateApplyProject(repoDir string,
 	return "", nil
 }
 
+func (a *DefaultCommandRequirementHandler) ValidateProjectDependencies(ctx command.ProjectContext) (failure string, err error) {
```
`Dependencies` makes sense here for the function name. I know I mentioned renaming some of the variables to `DependsOn`. I'm not going to nit pick on every single one, just use good judgment on which ones make sense to change.

GenPage reviewed (changes requested):
Great work. Thank you for your contribution. Please address nitpick comments. I envision we can deprecate `execution_order_group` after this. It's a bit ambiguous. I feel like overloading `depends_on` is a better place. If you need the projects executed in a particular order, that implies dependency. 

I can see needing only "ordering" vs a "hard fail" like dependencies. How difficult would it be to add a key somewhat like:

```yaml
version: 3
    projects:
    - dir: .
      name: project-2
      depends_on: 
           softFail: false
           projects: [project-1]
```

Thoughts?

nitrocode commented:
@GenPage 

> I envision we can deprecate execution_order_group after this

From @Luay-Sol 's above comment, it seems like we can use both execution order group and depends_on together, no?

Luay-Sol commented:
> Great work. Thank you for your contribution. Please address nitpick comments. I envision we can deprecate `execution_order_group` after this. It's a bit ambiguous. I feel like overloading `depends_on` is a better place. If you need the projects executed in a particular order, that implies dependency.
> 
> I can see needing only "ordering" vs a "hard fail" like dependencies. How difficult would it be to add a key somewhat like:
> 
> ```yaml
> version: 3
>     projects:
>     - dir: .
>       name: project-2
>       depends_on: 
>            softFail: false
>            projects: [project-1]
> ```
> 
> Thoughts?

Thanks for your feedback @GenPage. For your first point about deprecating `execution_order_group`, I agree and I think in the next iteration we can beef up `depends_on` and have it handle both cases (explicit project/workspace apply & global apply).

For your second point, It is not complicated to add such a key. However, I would like to better understand the use case. I am trying to figure out when a softfail will be used. If it is just to prevent failure, maybe not define any dependencies to the project.  

Thanks.

Luay-Sol commented:
> @GenPage
> 
> > I envision we can deprecate execution_order_group after this
> 
> From @Luay-Sol 's above comment, it seems like we can use both execution order group and depends_on together, no?

@nitrocode While it is true there's no issue with having both, I see merit points to merging the two. At the moment it is confusing to have 2 features that fundamentally try to do the same thing, i.e create an execution order, but each one handles a different use case, `explicit project/workspace apply` vs `global apply`.

The way I envision this is if we get the depend_on to do both, it will be much simpler for the user, to define it only once.

nitrocode commented:
ah ok, ty for the clarification.

jamengual commented:
> For your second point, It is not complicated to add such a key. However, I would like to better understand the use case. I am trying to figure out when a softfail will be used. If it is just to prevent failure, maybe not define any dependencies to the project.

I think @GenPage is referring to this PR use case? https://github.com/runatlantis/atlantis/pull/3323

GenPage commented:
> Thanks for your feedback @GenPage. For your first point about deprecating `execution_order_group`, I agree and I think in the next iteration we can beef up `depends_on` and have it handle both cases (explicit project/workspace apply & global apply).

@Luay-Sol Sounds good.

> For your second point, It is not complicated to add such a key. However, I would like to better understand the use case. I am trying to figure out when a softfail will be used. If it is just to prevent failure, maybe not define any dependencies to the project.

The idea is you would still get ordering, but the project doesn't stop the continuation of dependency execution. If we were to deprecate `execution_order_group` that functionality that provides an ordered run without stopping would be gone.

e.g. "Hey, I want to execute in this order but I don't care if this project fails to apply, I want to rest of the dependencies to run"

We can address that in a separate PR, where-as this PR just adds `depends_on`

GenPage commented:
@Luay-Sol Any update on addressing some of the code comments?

Luay-Sol commented:
> @Luay-Sol Any update on addressing some of the code comments?

@GenPage I will be getting back to this PR soon.

jamengual commented:
@Luay-Sol any chance you will get back to work on this?

Luay-Sol commented:
> @Luay-Sol any chance you will get back to work on this?

@jamengual I will be looking into it this week.

Luay-Sol commented:
FYI, find a problem with the current implementation related to workspaces with no plan, will still need to be applied if it is defined as a dependency.. looking into fixing that as well.

jamengual commented:
@Luay-Sol do you think you will have time to finish this?

jamengual commented:
Friendly ping @Luay-Sol

vincentgna commented:
Seems PR has been stale since April. We would also benefit from this feature.

@GenPage / @jamengual  / @Luay-Sol  - I cherry picked commit and addressed PR comments in

- #3821

GenPage commented:
I merged #3821. Thank you @vincentgna for pushing this through and making it so @Luay-Sol was a co-author.

Luay-Sol commented:
@vincentgna @GenPage first I want to apologize for not getting back to this issue sooner, I've been a bit busy. Id prefer we reopen this PR. As I mentioned in an earlier comment, the current implementation has a potential issue in the case where there are multiple workspaces defined where only some have a plan, the dependency will get into a weird state where it will keep forcing the order to update the workspaces with no plans. I have a potential fix that I've been testing in our fork that I'm hoping I can finalize this long weekend.

vincentgna commented:
@Luay-Sol  - I agree, but you can create a new PR (smaller changes are better than long lived branches)

Luay-Sol commented:
I totally agree.. My point @vincentgna is we should not merge the change in any case, knowing it has the problem. That's the main reason why I kept this branch open, all the above comments are minor changes, but the current implementation is fundamentally wrong, we are basically introducing a bug.

vincentgna commented:
do you want me to create a revert PR?

I'm also working on a forked release to add functionality I need on top of `depends_on`

Luay-Sol commented:
I think for the main project, maybe better to revert.

If you want to use it you own fork for now, you can go for it but just be aware of the bug I have mentioned.

As mentioned on my end I will be trying to put a proper fix/implementation this weekend.

Thanks

vincentgna commented:
> I think for the main project, maybe better to revert.
> 
> If you want to use it you own fork for now, you can go for it but just be aware of the bug I have mentioned.
> 
> As mentioned on my end I will be trying to put a proper fix/implementation this weekend.
> 
> Thanks

in my opinion:
- The `depends_on` in itself is behind a feature flag as it's opt-in
- Ideally we have pre-release / rc to sus out bugs
- For the sake of collaboration, it would be less confusing if you create an issue with a description of the bug, a way to reproduce,  ...

I'll defer to @GenPage / @jamengual if they prefer to have a revert PR for this

GenPage commented:
Re-opened and reverted the other PR.

vincentgna commented:
@Luay-Sol 

this is my proposal to evolve the `depends_on` field from this PR (not sure if you had time to work on it, but I'll be implementing this on my own fork so I hope we could get this aligned)

https://github.com/runatlantis/atlantis/issues/3241#issuecomment-1752422980

Luay-Sol commented:
Hi @vincentgna
These looks great additions, I am currently working on putting the proper fix for the plan vs apply dependency. Should get something ready for review soon.

Luay-Sol commented:
@vincentgna @GenPage This PR is ready for a final review, I have fixed the issue related to the `NoChanges` plan dependency, added a few more unit tests, as addressed all the comments in this PR.

Luay-Sol commented:
> @vincentgna @GenPage @jamengual This PR is now ready for a final review. I have addressed the issue related to the NoChanges plan dependency, added a few more unit tests, and incorporated all the comments made on this PR.
> 
> @vincentgna, I've also adjusted the documentation based on one of your closed PRs and refined the wording accordingly. Thank you for pointing it out.
> 
> The fix for the issue took longer than expected. Initially, I explored an alternative logic approach, but after extensive debugging, I found a simpler solution right in front of me, using the `PlannedNoChangesPlanStatus` to break the dependencies chain.
> 
> Please take a look at the PR when you have a chance.

That new status was introduced recently (after I proposed my PR) as part of this PR: https://github.com/runatlantis/atlantis/pull/3378. Now, it makes sense why I initially took a different approach to achieve the same result. So, thanks to @chroju for making it easier and somehow helping to ship this feature 😄 .

vincentgna commented on runatlantis.io/docs/repo-level-atlantis-yaml.md:
```diff
@@ -281,6 +283,46 @@ in each group one by one.
 
+Execution order groups are useful when you have dependencies between projects. However, they are only applicable in the case where
+you initiate a global apply for all of your projects, i.e `atlantis apply`. If you initiate an apply on a single project, then the execution order groups are ignored.
+Thus, the depond_on key is more useful in this case. and can be used in conjunction with execution order groups.
```
```suggestion
Thus, the `depends_on` key is more useful in this case. and can be used in conjunction with execution order groups.
```

vincentgna commented on server/events/command_requirement_handler.go:
```diff
@@ -65,6 +69,20 @@ func (a *DefaultCommandRequirementHandler) ValidateApplyProject(repoDir string,
+		for _, project := range ctx.PullStatus.Projects {
+
+			if project.ProjectName == dependOnProject && project.Status != models.AppliedPlanStatus && project.Status != models.PlannedNoChangesPlanStatus {
+				return fmt.Sprintf("Can't apply your project unless you apply its dependencies: [%s]", project.ProjectName), nil
```
should we mention this error in the docs?

> Can't apply your project unless you apply its dependencies:

Luay-Sol commented:
@jamengual @GenPage @vincentgna can I get final review on this? Anything else needed from me or I can do to get it out? Thanks.

GenPage reviewed (approved):
@Luay-Sol Excellent work, thank you for taking the time to contribute this.

GenPage commented:
@Luay-Sol Are you able to resolve the merge conflicts locally? We cannot resolve them via the web UI. Can you rebase from the Atlantis main branch please?

glasser commented:
Hi, just to check my understanding — in earlier discussions on this PR, the idea that this replaces `execution_order_group` is discussed. In the actual version merged, `depends_on` *only* adds potential validation errors and does *not* affect ordering when you run plan or apply without a specific subdirectory, right?  So likely folks will want to use them in conjunction with each other?

Luay-Sol commented:
> Hi, just to check my understanding — in earlier discussions on this PR, the idea that this replaces `execution_order_group` is discussed. In the actual version merged, `depends_on` _only_ adds potential validation errors and does _not_ affect ordering when you run plan or apply without a specific subdirectory, right? So likely folks will want to use them in conjunction with each other?

@glasser yes, you understood it correctly. Here's an example of how I use it:

```
projects:
  - name: development
    dir: .
    autoplan:
      when_modified: ["*.tf", "vars/development.tfvars"]
    execution_order_group: 1
    workspace: development
    workflow: infra
  - name: rc
    dir: .
    autoplan:
      when_modified: ["*.tf", "vars/rc.tfvars"]
    depends_on: ["development"]
    execution_order_group: 2    
    workspace: stability-test
    workflow: infra
  - name: staging
    dir: .
    autoplan:
      when_modified: ["*.tf", "vars/staging.tfvars"]
    depends_on: ["rc"]
    execution_order_group: 3
    workspace: staging
    workflow: infra 
    ```

hesse-nina commented:
> > Hi, just to check my understanding — in earlier discussions on this PR, the idea that this replaces `execution_order_group` is discussed. In the actual version merged, `depends_on` _only_ adds potential validation errors and does _not_ affect ordering when you run plan or apply without a specific subdirectory, right? So likely folks will want to use them in conjunction with each other?
> 
> @glasser yes, you understood it correctly. Here's an example of how I use it:
> 
> ```
> projects:
>   - name: development
>     dir: .
>     autoplan:
>       when_modified: ["*.tf", "vars/development.tfvars"]
>     execution_order_group: 1
>     workspace: development
>     workflow: infra
>   - name: rc
>     dir: .
>     autoplan:
>       when_modified: ["*.tf", "vars/rc.tfvars"]
>     depends_on: ["development"]
>     execution_order_group: 2    
>     workspace: stability-test
>     workflow: infra
>   - name: staging
>     dir: .
>     autoplan:
>       when_modified: ["*.tf", "vars/staging.tfvars"]
>     depends_on: ["rc"]
>     execution_order_group: 3
>     workspace: staging
>     workflow: infra 
>     ```
> ```

@Luay-Sol, does this work for you when you run a plan 'atlantis apply'? I'm running into an error where I have to run apply three times in this case, since rc doesn't know development has finished applying by the time it starts applying. This user also experienced this error: https://luppeng.wordpress.com/2025/01/11/atlantis-repo-config-execution_order_group-and-depends_on-attributes-may-be-mutually-exclusive-in-some-cases/
